### PR TITLE
CSP-1473: PR Employer Outer API: Send Email to Provider on Permission Created/Updated/Removed

### PR DESCRIPTION
**Pull request by shaun-downey-education**
- Extension of the post permissions endpoint to send out specific notification based on whether or not permissions are being updated, deleted or added.

### CODE REVIEW

File: src/EmployerPR/SFA.DAS.EmployerPR/Application/Commands/PostPermissions/PostPermissionsCommandHandler.cs [hunk 1]
```diff
@@ -1,12 +1,142 @@
...
+                }
+                break;
+            case HttpStatusCode.NotFound:
+                {
+                    existingOperations = [];
+                }
+                break;
+            default:
+                {
+                    throw new InvalidOperationException($"Invalid operation occurred trying to retrieve permissions for ukprn {command.Ukprn} and AccountLegalEntityId {command.AccountLegalEntityId}");
```
Comment (DevenKShah) [inline src/EmployerPR/SFA.DAS.EmployerPR/Application/Commands/PostPermissions/PostPermissionsCommandHandler.cs:43]: `Unexpected response code {permissionsResponse.ResponseMessage.StatusCode} returned when trying to ...`

File: src/EmployerPR/SFA.DAS.EmployerPR/Application/Commands/PostPermissions/PostPermissionsCommandHandler.cs [hunk 2]
```diff
@@ -1,12 +1,142 @@
...
+
+        return Unit.Value;
+    }
+
+    private static bool NoUpdatesRequired(List<Operation> requestedPermissions, List<Operation> currentPermissions)
+    {
+        return (requestedPermissions.Count == 0 && currentPermissions.Count == 0) || IdenticalPermissions(requestedPermissions, currentPermissions);
+    }
+
+    private static bool IdenticalPermissions(List<Operation> requestedPermissions, List<Operation> currentPermissions)
```
Comment (DevenKShah) [inline src/EmployerPR/SFA.DAS.EmployerPR/Application/Commands/PostPermissions/PostPermissionsCommandHandler.cs:85]: booleans should read like questions `HasIdenticalPermissions` or `IdenticalPermissionsExists`

File: src/EmployerPR/SFA.DAS.EmployerPR/Application/Commands/PostPermissions/PostPermissionsCommandHandler.cs [hunk 3]
```diff
@@ -1,12 +1,142 @@
...
     {
-        return await _providerRelationshipsApiRestClient.PostPermissions(command, cancellationToken);
+        return currentPermissions.OrderBy(a => (int)a).SequenceEqual(requestedPermissions.OrderBy(a => (int)a));
+    }
+
+    private static PermissionEmailTemplateType GetTemplateType(List<Operation> operations)
```
Comment (DevenKShah) [inline src/EmployerPR/SFA.DAS.EmployerPR/Application/Commands/PostPermissions/PostPermissionsCommandHandler.cs:90]: keep the name as `existingOperations` so it continues to read correctly

File: src/EmployerPR/SFA.DAS.EmployerPR.UnitTests/Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandlerTests.cs [hunk 4]
```diff
@@ -1,32 +1,388 @@
...
+using System.Text.Json;
 
 namespace SFA.DAS.EmployerPR.UnitTests.Application.Permissions.Commands.PostPermissions;
 
 public class PostPermissionsCommandHandlerTests
 {
     [Test]
     [MoqAutoData]
-    public async Task Handle_Returns_PostPermissionsCommandResult(
+    public async Task PostPermissionsCommandHandler_Handle_Requires_No_Action(
```
Comment (DevenKShah) [inline src/EmployerPR/SFA.DAS.EmployerPR.UnitTests/Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandlerTests.cs:23]: Suggest method name change Handle_NoRequestedAndExistingOperations_NoAction

File: src/EmployerPR/SFA.DAS.EmployerPR.UnitTests/Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandlerTests.cs [hunk 5]
```diff
@@ -1,32 +1,388 @@
...
+            x.PostPermissions(
+                It.IsAny<PostPermissionsCommand>(),
+                It.IsAny<CancellationToken>()
+            ),
+        Times.Never);
+    }
+
+    [Test]
+    [MoqAutoData]
+    public void PostPermissionsCommandHandler_Handle_Throws_InvalidOperationException(
```
Comment (DevenKShah) [inline src/EmployerPR/SFA.DAS.EmployerPR.UnitTests/Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandlerTests.cs:75]: Handle_UnexpectedResponseOnExistingPermission_ThrowsException

File: src/EmployerPR/SFA.DAS.EmployerPR.UnitTests/Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandlerTests.cs [hunk 6]
```diff
@@ -1,32 +1,388 @@
...
+using System.Net;
+using System.Text.Json;
 
 namespace SFA.DAS.EmployerPR.UnitTests.Application.Permissions.Commands.PostPermissions;
 
 public class PostPermissionsCommandHandlerTests
 {
     [Test]
     [MoqAutoData]
-    public async Task Handle_Returns_PostPermissionsCommandResult(
```
Comment (DevenKShah) [inline src/EmployerPR/SFA.DAS.EmployerPR.UnitTests/Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandlerTests.cs:15]: A test case for identical operations resulting in no action is not covered

File: src/EmployerPR/SFA.DAS.EmployerPR.UnitTests/Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandlerTests.cs [hunk 7]
```diff
@@ -1,32 +1,388 @@
...
-                command,
+                It.IsAny<PostPermissionsCommand>(),
+                It.IsAny<CancellationToken>()
+            ),
+        Times.Never);
+    }
+
+    [Test]
+    [MoqAutoData]
+    public async Task PostPermissionsCommandHandler_Handle_Requires_Permissions_Removed(
```
Comment (DevenKShah) [inline src/EmployerPR/SFA.DAS.EmployerPR.UnitTests/Application/Permissions/Commands/PostPermissions/PostPermissionsCommandHandlerTests.cs:119]: Handle_NoOperationsRequestedWithExistingPermissions_RemovesExistingPermissions

**Review by DevenKShah — COMMENTED** [review]
Looks good mostly, test names can be improved, I have added some suggestions. Also one test case is missing.

**Comment by shaun-downey-education** [comment]
Requires rebase on 1471

**Pull request closed without merging**